### PR TITLE
minimega: vnc api extensions

### PR DESCRIPTION
**Pull request by cdshann**
This is a refactor of the minimega vnc recording and playback code to support the features requested in #426.

This is not yet ready to merge, but I wanted to create the PR to get any feedback and so everyone could track the development.

### CODE REVIEW

File: src/minimega/vnc_cli.go [hunk 1]
```diff
@@ -9,19 +9,20 @@ import (
...
 	"strconv"
+	"time"
 )
 
 var vncCLIHandlers = []minicli.Handler{
 	{ // vnc
-		HelpShort: "record or playback VNC kb or fb",
+		HelpShort: "record VNC kb or fb",
 		HelpLong: `
 Record or playback keyboard and mouse events sent via the web interface to the
```
Comment (jcrussell) [inline src/minimega/vnc_cli.go]: Remove `or playback`.

File: src/minimega/vnc_cli.go [hunk 2]
```diff
@@ -9,19 +9,20 @@ import (
...
 selected VM. Can also record the framebuffer for the specified VM so that a
 users can watch a video of interactions with the VM.
 
 With no arguments, vnc will list currently recording or playing VNC sessions.
 
-If record is selected, a file will be created containing a record of mouse and
+Ifr reecord is selected, a file will be created containing a record of mouse and
```
Comment (jcrussell) [inline src/minimega/vnc_cli.go]: Spelling.

File: src/minimega/vnc_cli.go [hunk 3]
```diff
@@ -30,11 +31,23 @@ to the specified VM.`,
...
-			"vnc <noplayback,> <host> <vm id or name>",
+			"vnc <kb,fb> <stop,> <host> <vm id or name>",
 		},
-		Call: wrapSimpleCLI(cliVNC),
+		Call: wrapSimpleCLI(cliVNCRecord),
+	},
+	{ // play vnc kb
+		HelpShort: "play VNC kb",
+		HelpLong: `
+Play a previously recorded VNC kb`,
```
Comment (jcrussell) [inline src/minimega/vnc_cli.go]: Document what each of these do.

**Comment by jcrussell** [comment]
We should eliminate the `<host>` argument from these commands. With namespaces, who knows where a VM may land. Instead, use `globalVMs` to find a VM with a matching name and set the host from the responses. This also makes it easier to support #342 because we'll be able to lookup the VNC port at the same time.

If the user isn't using namespaces and they are running VMs with the same name... perhaps we could support syntax like `<host>:<vm>`.

**Comment by cdshann** [comment]
This is just an update on the features implemented so far in this PR:

- Listing running VNC recordings will display the elapsed time, running playbacks will show the time remaining
- You can now start, stop, pause, and continue playbacks. 
- Added the step command to immediately move to the next line in a playback file
- Getstep returns the current position in the playback
- Users can embed comments in VNC playbacks which are logged at the info level
- Users can immediately inject arbitrary VNC commands from the command line with the new inject command

Before this is merged I need to make some changes so that this integrates with namespaces.

**Comment by mkunz7** [comment]
Any status update on this? Would love to start using it.

**Comment by djfritz** [comment]
@cdshann what's the status on this?

**Comment by jcrussell** [comment]
Next steps:

 * Move recording to IP <-> domain socket
 * Remove `host` parameter -- should use `wrapVMTargetCLI`

**Comment by cdshann** [comment]
Now that #676 was merged I am working on this PR again. Right now the API is:
```
vnc <kb,fb> <record,> <vm name> <filename>
vnc <kb,fb> <stop,> <vm name>
vnc <play,> <host> <vm id or name> <filename>
vnc <stop,> <host> <vm id or name>
vnc <pause,> <host> <vm id or name>
vnc <continue,> <host> <vm id or name>
vnc <step,> <host> <vm id or name>
vnc <getstep,> <host> <vm id or name>
vnc <inject,> <host> <vm id or name> <cmd>
```
Any API changes and/or (smaller) features you would like to see? @djfritz @jcrussell @floren @mkunz7

**Comment by jcrussell** [comment]
I'd like to swap the order of `<kb,fb>` and `<record>` -- seems more logical to me.

Also, none of those commands should include a host, right?

File: src/minimega/vnc_cli.go [hunk 4]
```diff
@@ -57,90 +83,145 @@ List all running vnc playback/recording instances. See "help vnc" for more infor
...
-				hostname, v.VM.Name, strconv.Itoa(v.VM.ID),
-				"record kb",
-				v.file.Name(),
-			})
-		}
+	fname := c.StringArgs["filename"]
+
+	vm, err := vms.FindKvmVM(c.StringArgs["vm"])
+	if err != nil {
+		return fmt.Errorf("vm %s not found", c.StringArgs["vm"])
```
Comment (jcrussell) [inline src/minimega/vnc_cli.go]: Return the err here, don't create a new one.

File: src/minimega/vnc_cli.go [hunk 5]
```diff
@@ -57,90 +83,145 @@ List all running vnc playback/recording instances. See "help vnc" for more infor
...
 			return client.Stop()
 		}
+	}
+	return err
+}
+
+// List all active recordings and playbacks
+func cliVNCList(c *minicli.Command, resp *minicli.Response) error {
+	resp.Header = []string{"name", "type", "time", "filename"}
+	resp.Tabular = [][]string{}
```
Comment (jcrussell) [inline src/minimega/vnc_cli.go]: Can delete this line.

File: src/minimega/vnc_record.go [hunk 6]
```diff
@@ -0,0 +1,229 @@
...
+var (
+	vncKBRecording     map[string]*vncKBRecord
+	vncFBRecording     map[string]*vncFBRecord
+	vncKBRecordingLock sync.RWMutex
+	vncFBRecordingLock sync.RWMutex
+)
+
+func init() {
+	vncKBRecording = make(map[string]*vncKBRecord)
+	vncFBRecording = make(map[string]*vncFBRecord)
```
Comment (jcrussell) [inline src/minimega/vnc_record.go]: Can move these makes up into the var.

**Comment by cdshann** [comment]
Stress tested this I wasn't able to find any concurrency issues, and the memory usage is aligned with the current VNC implementation.

Should be good to merge @jcrussell @djfritz

**Comment by djfritz** [comment]
LGTM, is it worth adding an article on the vnc capability?

**Pull request merged**